src/adapters/sqlite/makeDispatcher/node/DatabaseDriver.js

```diff
@@ -0,0 +1,292 @@
+    return results.map((row: any) => {
+      const id = `${row.id}`
+      if (this.isCached(table, id)) {
+        return id
+      }
+      this.markAsCached(table, id)
+      return row
+    })
+  }
+
+  query = (table: string, query: string) => this.cachedQuery(table, query)
+
+  count = (query: string) => this.database.count(query)
+
+  batch = (operations: any[]) => {
+    const newIds = []
+    const removedIds = []
+
+    this.database.inTransaction(() => {
+      operations.forEach((operation: any[]) => {
+        const [type, table, ...rest] = operation
+        switch (type) {
+          case 'execute': {
+            const [query, args] = rest
+            this.database.execute(query, fixArgs(args))
+            break
+          }
+
+          case 'create': {
+            const [id, query, args] = rest
+            this.database.execute(query, fixArgs(args))
+            newIds.push([table, id])
+            break
+          }
+
+          case 'markAsDeleted': {
+            const [id] = rest
+            const query = `UPDATE '${table}' SET _status='deleted' WHERE id == ?`
+            this.database.execute(query, [id])
+            removedIds.push([table, id])
+            break
+          }
+
+          case 'destroyPermanently': {
+            const [id] = rest
+            // TODO: What's the behavior if nothing got deleted?
+            this.database.execute(`DELETE FROM '${table}' WHERE id == ?`, [id])
+            removedIds.push([table, id])
+            break
+          }
+
+          default: {
+            throw new Error('unreachable')
+          }
+        }
+      })
+    })
+
+    newIds.forEach(([table, id]) => {
+      this.markAsCached(table, id)
+    })
+
+    removedIds.forEach(([table, id]) => {
+      this.removeFromCache(table, id)
+    })
+  }
+
+  getDeletedRecords = (table: string): string[] => {
+    return this.database
+      .queryRaw(`SELECT ID FROM '${table}' WHERE _status='deleted'`)
+      .map(row => `${row.id}`)
+  }
+
+  destroyDeletedRecords = (table: string, records: string[]) => {
+    const recordPlaceholders = records.map(() => '?').join(',')
+    this.database.execute(`DELETE FROM '${table}' WHERE id IN (${recordPlaceholders})`, records)
+  }
+
+  // MARK: - LocalStorage
+
+  getLocal = (key: string) => {
+    const results = this.database.queryRaw('SELECT `value` FROM `local_storage` WHERE `key` = ?', [
+      key,
+    ])
+
+    if (results.length > 0) {
+      return results[0].value
+    }
+
+    return null
+  }
+
+  setLocal = (key: string, value: any) => {
+    this.database.execute('INSERT OR REPLACE INTO `local_storage` (key, value) VALUES (?, ?)', [
+      key,
+      `${value}`,
+    ])
+  }
+
+  removeLocal = (key: string) => {
+    this.database.execute('DELETE FROM `local_storage` WHERE `key` == ?', [key])
+  }
+
+  // MARK: - Record caching
+
+  hasCachedTable = (table: string) =>
+    Object.prototype.hasOwnProperty.call(this.cachedRecords, table)
```

Comment:
it's OK, but it would be better to use a Map for this to avoid having to do this dance

Reply:
I'm trying to remember WHY I did this instead of `map`. I personally hate this approach.

Reply:
yeah, me too. Last time I checked, it wasn't even faster in microbenchmarks